programs/fileio_asyncio.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) Yann Collet, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under both the BSD-style license (found in the
+ * LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ * in the COPYING file in the root directory of this source tree).
+ * You may select, at your option, one of the above-listed licenses.
+ */
+
+#ifndef ZSTD_FILEIO_ASYNCIO_H
+#define ZSTD_FILEIO_ASYNCIO_H
+
+#if defined (__cplusplus)
+extern "C" {
+#endif
+
+#include "../lib/common/mem.h"     /* U32, U64 */
+#include "fileio_types.h"
+#include "platform.h"
+#include "util.h"
+#include "../lib/common/pool.h"
+#include "../lib/common/threading.h"
+
+#define MAX_IO_JOBS          (10)
+
+typedef struct {
+    /* These struct fields should be set only on creation and not changed afterwards */
+    POOL_ctx* threadPool;
+    int totalIoJobs;
+    FIO_prefs_t* prefs;
+    POOL_function poolFunction;
+
+    /* Controls the file we currently write to, make changes only by using provided utility functions */
+    FILE* file;
+
+    /* The jobs and availableJobsCount fields are accessed by both the main and worker threads and should
+     * only be mutated after locking the mutex */
+    ZSTD_pthread_mutex_t ioJobsMutex;
+    void* availableJobs[MAX_IO_JOBS];
+    int availableJobsCount;
+} io_pool_ctx_t;
+
+typedef struct {
+    io_pool_ctx_t base;
+    unsigned storedSkips;
+} write_pool_ctx_t;
+
+typedef struct {
+    /* These fields are automatically set and shouldn't be changed by non WritePool code. */
+    void *ctx;
+    FILE* file;
+    void *buffer;
+    size_t bufferSize;
+
+    /* This field should be changed before a job is queued for execution and should contain the number
+     * of bytes to write from the buffer. */
+    size_t usedBufferSize;
+    U64 offset;
+} io_job_t;
+
+/** FIO_fwriteSparse() :
+*  @return : storedSkips,
+*            argument for next call to FIO_fwriteSparse() or FIO_fwriteSparseEnd() */
+unsigned FIO_fwriteSparse(FILE* file,
+                 const void* buffer, size_t bufferSize,
+                 const FIO_prefs_t* const prefs,
+                 unsigned storedSkips);
+
+void FIO_fwriteSparseEnd(const FIO_prefs_t* const prefs, FILE* file, unsigned storedSkips);
+
+/* WritePool_releaseIoJob:
+ * Releases an acquired job back to the pool. Doesn't execute the job. */
+void WritePool_releaseIoJob(io_job_t *job);
```

Comment:
minor : this choice of symbol prefix, `WritePool_`, contrasts with the rest of the code base.
There's no absolute "wrong" about it, it's just a matter of coding convention consistency.

Suggested alternative prefix : `WP_`.

Also : it's a bit strange to have some `FIO_` functions defined in different units.
In general, symbol prefix and units should be related to each other.
Maybe there is a need to pick a side here.

Reply:
Can definitely change `WritePool_` to `WP_` although I'd usually opt for more explicit names. Note that in the next patch I'll also introduce `ReadPool_` functions. So the two options I see are to use `WP_` and `RP_` or `AIO_writePool_` and `AIO_readPool_`.

As for `FIO_fwriteSparse*` -- I can add the same prefix to them as well, shouldn't be a problem.  Note that my next PR will turn them into internal (static) functions for the `fileio_asyncio` module, and they will be removed from the header, so possibly just having them with no prefix is good enough.

Reply:
@Cyan4973 -  Would you prefer `AIO_writePool_` or `WP_` (next patch would introduce `AIO_readPool_` or `RP_` respectively)?
My recommendation would be `AIO_writePool_`.

Reply:
It depends how much you value brevity.
Both look fine to me, hence `AIO_writePool_` is fine.

Also, subtle thing, if `WritePool` is itself an object / type, then you can use `AIO_WritePool_`, since it refers directly to the type being manipulated (in which case, I would expect a `WritePool*` pointer as first argument).